### PR TITLE
feat : add studio tour for first time user

### LINKED ISSUE #284: Create onboarding for features of Studio
#### Reason/Context
I think the Studio could use an onboarding process where there are a few steps for new users to go through to get started in Studio, especially if coming from Playground. I think it could be a modal with 3-5 simple steps to getting started, perhaps explaining that you can start from a template, show the visualizer, etc. I think we could also use highlighting at each step in the area where the user would take an action.

I don't feel like this is a super urgent thing to work on, but thought of it when reviewing PR #282 where I was thinking of ways that we could reduce the amount of text content in the "New features" modal.

**Some examples for inspiration:**
- Dropbox has a good example for what I am thinking about -> [See this article
](https://uxplanet.org/3-awesome-user-onboarding-flows-for-web-c8b1ec6a508a)

### PR DESCRIPTION
**Pull request by aialok**
**Description**

- This PR will introduce a tour for first-time users.
- There are a total of ten steps explaining each component of the studio.

**Related issue(s)**
- Closes #284 

Here's is snapshot : 

[Screencast from 2024-06-27 22-55-36.webm](https://github.com/asyncapi/studio/assets/66772290/dce842ac-9264-4104-89b7-78b3270f1c47)

### CODE REVIEW

**Comment by aialok** [comment]
This PR is ready to review.
Thank you  : )

**Comment by Shurtu-gal** [comment]
Is it just me or nobody is able to see the tour?

**Comment by Shurtu-gal** [comment]
@aialok the changes need to be done in `studio-next` as well I suppose. @Amzani when are we planning to merge these two?

Also is there any scope of showing interaction to user during the tour.

**Comment by aeworxet** [comment]
@asyncapi/bounty_team

**Comment by Shurtu-gal** [comment]
**Updates:**

I have told @aialok to do a comparison b/w various option such as [react-joyride](https://react-joyride.com/controlled) and create an `ADR` so that we can make an informed decision. The only problem with the current implementation is the lack of interactivity (controlled elements).

**Comment by aialok** [comment]
@Mayaleeeee Please review the design of the popover.
@Shurtu-gal This is ready for review.

Thank you : )

**Comment by Shurtu-gal** [comment]
Great work @aialok 🚀.
Just need a design approval, meanwhile this can be put on slack as well, for suggestions from everyone, can cc `@studio` there as well.

File: apps/studio-next/src/helpers/driver.ts
Comment (Amzani) [inline apps/studio-next/src/helpers/driver.ts:17]:
```suggestion
          "Discover a powerful tool for designing, documenting, and managing AsyncAPI-based applications. This tour will guide you through key features to enhance your AsyncAPI development workflow.",
```

File: apps/studio-next/src/helpers/driver.ts
Comment (Amzani) [inline apps/studio-next/src/helpers/driver.ts:25]: API -> AsyncAPI

**Review by Amzani — COMMENTED** [review]
Great PR, that will improve our users onboarding!
In addition to my review 

The close button could be improved by removing the hover
Screenshot 2024-07-23 at 10 19 16

**Comment by aialok** [comment]
Done with all the requested changes.
Thank you @Amzani : )

File: apps/studio-next/src/helpers/driver.ts
Comment (KhudaDad414) [inline apps/studio-next/src/helpers/driver.ts:35]:
```suggestion
          "Explore your API structure using this information panel. Quickly access Servers, Channels, Operations, Messages, and Schemas - the building blocks of your AsyncAPI document.",
```

File: apps/studio-next/src/helpers/driver.ts
Comment (KhudaDad414) [inline apps/studio-next/src/helpers/driver.ts:33]: IMO it's better to call this one `Navigation Panel` and adjust the id's accordingly. It helps user navigate the AsyncAPI document right?

File: apps/studio-next/src/helpers/driver.ts
Comment (KhudaDad414) [inline apps/studio-next/src/helpers/driver.ts:35]:
```suggestion
          "Explore your API structure using this navigation panel. Quickly access Servers, Channels, Operations, Messages, and Schemas - the building blocks of your AsyncAPI specification.",
```

**Review by KhudaDad414 — COMMENTED** [review]
Looks great, left a few suggestions.

File: apps/studio-next/src/helpers/driver.ts
Comment (aialok) [reply apps/studio-next/src/helpers/driver.ts]: Yes, right.

**Comment by aialok** [comment]
Thank you @KhudaDad414  , done with requested changes : )

File: apps/studio/src/components/Sidebar.tsx
Comment (Shurtu-gal) [inline apps/studio/src/components/Sidebar.tsx:142]: Can we store the current tour step so that user doesn't have to go through everything again?

Like this `driverObj.drive(4); // Starts at step 4`

**Review by Shurtu-gal — CHANGES_REQUESTED** [review]

File: apps/studio/src/components/Sidebar.tsx
Comment (aialok) [reply apps/studio/src/components/Sidebar.tsx:142]: Yes, we can store the state of steps in local storage.

But there might be one problem: what if a user comes for the first time, checks 2-3 steps, and then closes the tour because they don't want to continue? When they come back, the tour will start from where they left off, this might be bad experience for the user.

Currently, the tour is only going to run for the first-time visit to the studio; after that, it won't appear because we are storing whether the user has visited or not in local storage.

We have also provided a button for the user to restart the tour at any time.

So, should I proceed with your given approach but there might be some problem with user experience?

Comment (Shurtu-gal) [reply apps/studio/src/components/Sidebar.tsx:142]: What I meant is if user clicks the start tour button, in that case this can done I think.

Comment (aialok) [reply apps/studio/src/components/Sidebar.tsx:142]: My bad, I thought of first time user part. 
Sure, we can implement this.

So, should I add this part when user restart to tour again?

Comment (Shurtu-gal) [reply apps/studio/src/components/Sidebar.tsx:142]: I think so. WDYT @Amzani @KhudaDad414

Comment (aialok) [reply apps/studio/src/components/Sidebar.tsx:142]: @Shurtu-gal Is there anything to add more?

Comment (Shurtu-gal) [reply apps/studio/src/components/Sidebar.tsx:142]: Is this done ?

Comment (aialok) [reply apps/studio/src/components/Sidebar.tsx:142]: No, not yet waiting for the reply of other two people : )
Should I go for it?

Comment (Shurtu-gal) [reply apps/studio/src/components/Sidebar.tsx:142]: I think you should 👍🏻

Comment (aialok) [reply apps/studio/src/components/Sidebar.tsx]: Done.
Added feature to resume tour from last position.

[Added feature to resume tour from last position.](https://github.com/user-attachments/assets/017b7e40-1cdf-48ea-be85-29a122ce0fa8)

Thank you @Shurtu-gal for recommending this amazing feature : )

**Review by Shurtu-gal — APPROVED** [review]
LGTM 🚀.

Thanks for the work @aialok.

**Comment by KhudaDad414** [comment]
@aialok the branch has some conflicts.

**Comment by aialok** [comment]
> @aialok the branch has some conflicts.

Oops. It pnpm-lock file conflict it is lil hard to resolve but lemme fix it : )

**Comment by aialok** [comment]
Thank you @KhudaDad414 
I have resolved the conflict  : )

**Comment by Shurtu-gal** [comment]
@Amzani I suppose the sonarcloud check for duplication was disabled?

**Comment by aialok** [comment]
Can we merge this PR? or is there anything else need to be done?
Thank you  : )

**Comment by Amzani** [comment]
Some tests are still failing

**Comment by aialok** [comment]
> Some tests are still failing

Fixing it. It is simple eslint error : )

**Comment by aialok** [comment]
@Amzani Fixed the lint error.
Hopefully test will pass now : )

**Comment by aeworxet** [comment]
@aialok
Please also pay attention to the [sonarcloud analysis](https://github.com/asyncapi/studio/pull/1120#issuecomment-2278035912).

**Comment by aialok** [comment]
> @aialok
> Please also pay attention to the [sonarcloud analysis](https://github.com/asyncapi/studio/pull/1120#issuecomment-2278035912).

Basically we have same code in two monorepo one in studio and studio-next.

**Comment by aeworxet** [comment]
FYI

Duplication reported by SonarCloud are two mostly identical files:
```
./apps/studio/src/helpers/driver.ts
./apps/studio-next/src/helpers/driver.ts
```

Maintainers, should they both be present in one PR or should there be two PRs, separate for `studio` and `studio-next`, to keep SonarCloud happy?

**Comment by aialok** [comment]
> FYI
> 
> Duplication reported by SonarCloud are two mostly identical files:
> 
> ```
> ./apps/studio/src/helpers/driver.ts
> ./apps/studio-next/src/helpers/driver.ts
> ```
> 
> Maintainers, should they both be present in one PR or should there be two PRs, separate for `studio` and `studio-next`, to keep SonarCloud happy?

https://github.com/asyncapi/studio/pull/1120#issuecomment-2198520537

**Comment by Shurtu-gal** [comment]
@aeworxet these are actually supposed be merged by now. Ref: https://github.com/asyncapi/studio/pull/1120#issuecomment-2198520537

I would suggest merging this PR and then removing the legacy `studio` folder.

**Comment by aialok** [comment]
@Amzani Can you please take a look at this PR? 
This is ready to merge.
Thank you : )

**Comment by aeworxet** [comment]
@fmvilas, is there a backup person who can also approve this functionality for merging and be contacted about future ones? We're having a bit of a bottleneck here due to @Amzani being the only one.

**Comment by aeworxet** [comment]
Yep. Great. Good rest is what people need to do.

image

**Comment by aeworxet** [comment]
@magicmatatjahu, can you please take a look at this fabulous PR?

Would you approve it for merging?

**Review by Amzani — DISMISSED** [review]
LGTM

**Comment by aeworxet** [comment]
> /rtm

Well, it was worth trying.

**Comment by derberg** [comment]
instead of disabling sonar, I recommend you add the problematic file to https://github.com/asyncapi/studio/blob/master/.sonarcloud.properties with explanation why code duplication is ok in this case.

also, I'm not sure how things work in this repo, but shouldn't you have a changeset file to have that released? - asking cause I see `feat` in title that suggest it is a new feature (even if you do not use semantic release, conventional commits still have a meaning)

**Comment by aeworxet** [comment]
@Amzani
As @derberg [says](https://github.com/asyncapi/studio/pull/1120#issuecomment-2338425864), should this PR have a changeset to bump the minor version?

@Shurtu-gal @aialok
How about specifying [problematic files](https://github.com/asyncapi/studio/pull/1120#issuecomment-2280182791) in the `.sonarcloud.properties` indeed? Because even though `studio-next` is supposed to replace `studio` in the future, two versions simultaneously will still be there for some time.

**Comment by Shurtu-gal** [comment]
> @Shurtu-gal @aialok How about specifying [problematic files](https://github.com/asyncapi/studio/pull/1120#issuecomment-2280182791) in the `.sonarcloud.properties` indeed? Because even though `studio-next` is supposed to replace `studio` in the future, two versions simultaneously will still be there for some time.

Yeah this can indeed be done. @aialok will you be doing this or should I do it on my end?

**Comment by aialok** [comment]
@Shurtu-gal please do at your end.
I am little busy with some other stuff.

Thank you 🙌

**Comment by Shurtu-gal** [comment]
> @Shurtu-gal please do at your end. I am little busy with some other stuff.
> 
> Thank you 🙌

Just saw that as I am not a CODEOWNER so I don't have access to your branch 😉. Could you give me the access or maybe just add the folder for ignoring under `sonar.cpd.exclusions`.

**Comment by aialok** [comment]
@Shurtu-gal Okay lemme do it  : )
which are the folders need to put in `sonar.cpd.exclusions`?

**Comment by Shurtu-gal** [comment]
> @Shurtu-gal Okay lemme do it : ) which folders need to put in `sonar.cpd.exclusions`?

All the folder and files having duplication issue.

**Comment by aeworxet** [comment]
Asked @Amzani to add @Shurtu-gal as a code owner to accelerate the decision-making process.

@Shurtu-gal, merge this PR at last, please.

**Comment by Shurtu-gal** [comment]
Sure, @Amzani has told that he will do it by Friday this week.

**Comment by Shurtu-gal** [comment]
@Amzani could you approve the workflow run.

**Comment by aeworxet** [comment]
> @Amzani could you approve the workflow run.

@Amzani
Should @Shurtu-gal be granted some additional permissions in the repository's settings?

**Comment by aeworxet** [comment]
https://github.com/asyncapi/studio/pull/1094 was merged, so pay attention to this change: https://github.com/asyncapi/studio/pull/1094#issuecomment-2344700719.

**Comment by aeworxet** [comment]
[Hard-locked](https://github.com/asyncapi/studio/pull/1139) the `@asyncapi/converter` on version `1.5.0` during handling incident https://github.com/asyncapi/studio/issues/1138.
Just a heads-up for conflict resolution.

**Comment by aialok** [comment]
> [Hard-locked](https://github.com/asyncapi/studio/pull/1139) the `@asyncapi/converter` on version `1.5.0` during handling incident https://github.com/asyncapi/studio/issues/1138.
> Just a heads-up for conflict resolution.


Sure..sure...I will be resolving the confict by night.

Thanks @aeworxet

**Comment by aialok** [comment]
> [Hard-locked](https://github.com/asyncapi/studio/pull/1139) the `@asyncapi/converter` on version `1.5.0` during handling incident #1138. Just a heads-up for conflict resolution.

Sure..Sure...conflict in pkg json and lock file...little hard to

**Comment by aialok** [comment]
@Shurtu-gal might be because there is some conflict in the lock file...I just resolve the conflict.

Can I delete lock file and generate new one and push it?

**Comment by Shurtu-gal** [comment]
Yeah, that would be great 😃

**Comment by aialok** [comment]
@Shurtu-gal  I have resolved all the conflict. Please merge it. It been so long : )
Thank you  : )

**Comment by Shurtu-gal** [comment]
@Amzani could you please approve the workflows and approve this. 
Thanks 🙏

**Comment by aialok** [comment]
Fixing Build failed...hold on : )

**Comment by Shurtu-gal** [comment]
Also @aialok could you run `yarn lint` as well.

**Comment by aialok** [comment]
> Also @aialok could you run `yarn lint` as well.

Fixes the build : )

**Comment by aeworxet** [comment]
> @aialok Sonar [still complains](https://sonarcloud.io/project/issues?id=asyncapi_studio&pullRequest=1120&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true), this time it's justified.

https://github.com/asyncapi/studio/pull/1120#issuecomment-2356991212

**Comment by aialok** [comment]
@aeworxet fixing it : )

**Review by Amzani — APPROVED** [review]
LGTM

**Comment by aialok** [comment]
@Shurtu-gal It got approved now. Can we go with the merge?

**Review by Shurtu-gal — APPROVED** [review]

**Comment by Shurtu-gal** [comment]
> @Shurtu-gal It got approved now. Can we go with the merge?

Absolutely, mate 🚀. Great work!!

**Comment by aialok** [comment]
> > @Shurtu-gal It got approved now. Can we go with the merge?
> 
> Absolutely, mate 🚀. Great work!!

It was really fun to work with you ❤️❤️

**Pull request merged**
